src/pod.ts

```diff
@@ -362,19 +420,116 @@ export interface AbstractPodProps extends base.ResourceProps {
  */
 export interface PodProps extends AbstractPodProps {}
 
+/**
+ * Match a resource by labels.
+ */
+export class LabelSelector {
+
+  /**
+   * Create a selector from label queries.
+   */
+  public static of(...queries: LabelQuery[]): LabelSelector {
+    return new LabelSelector(queries);
+  }
+
+  private constructor(public readonly queries: LabelQuery[]) {}
+
+  /**
+   * @internal
+   */
+  public _toKube(): k8s.LabelSelector {
+    if (this.queries.length === 0) return {};
+    return { matchExpressions: this.queries.map(q => ({ key: q.key, operator: q.operator, values: q.values })) };
+  }
+}
+
+/**
+ * Represents an object that can select pods.
+ */
+export interface IPodSelector {
+  /**
+   * Return the label selector that selects the pods.
+   *
+   * - Implementors should return `Pod.all()` to select all pods.
+   */
+  toPodLabelSelector(): LabelSelector;
+}
+
+/**
+ * Represents an object that can select pods in particular namespaces.
+ */
+export interface INamespacedPodSelector {
+  /**
+   * Return the pod selector that selects the pods.
+   */
+  toPodSelector(): IPodSelector;
+
+  /**
+   * Return the label selector that selects the namespaces.
+   *
+   * - Implementors should return `undefined` to select "this" namespace (i.e the namespace of the selected pod).
+   * - Implementors should return `Namespace.all()` to select all namespaces.
+   */
+  toNamespaceSelector(): namespace.INamespaceSelector | undefined;
+}
```

Comment:
Could we simplify this to something like `INamespacedPodSelector extends IPodSelector, INamespaceSelector`?

Also -- from my reading of this API it looks like the user can select pods by their labels or which namespace they're in, but not necessarily by the node they're in. Can you check if my intuition on that is right? There's a lot of "selector"-ish concepts being introduced in this PR so I might be totally wrong.

Reply:
As discussed internally, having `INamespacedPodSelector extend IPodSelector` will mean that `Pod.labeled(...).namespaced(...)` can be passed to resources expecting only `IPodSelector`.

That is, some resources don't allow namespace selection in their selector definition, and those resources simply accept an `IPodSelector`. It shouldn't be possible to pass `INamespacedPodSelector` to those resources.

Reply:
> Also -- from my reading of this API it looks like the user can select pods by their labels or which namespace they're in, but not necessarily by the node they're in. Can you check if my intuition on that is right? There's a lot of "selector"-ish concepts being introduced in this PR so I might be totally wrong.

Users can select pods by one of three methods:

1. Strict pod label matching (not modeled in this PR just yet)
2. Pod label queries (`IPodSelector`)
3. Pod label queries and namespaces (`INamespacedPodSelector`)

There is no API saying "select a pod that runs on this node". There's only "select a node with these labels, and run the pod on it".